pytorch: 1.0.0 -> 1.2.0

stites opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Closes #63073
Closes #65041
Closes #67468

This depends on @tbenst 's https://github.com/NixOS/nixpkgs/pull/61347 (thank you!). See #63073 and this PR thread for more details. 

Work can be previewed at [stites/pytorch-world](https://github.com/stites/pytorch-world). Building from source takes a long time, please check out the [hosted cachix binaries](https://pytorch-world.cachix.org/)

Changelog:
- updates to pytorch-1.2.0
- runs all tests except for `utils`
- adds a `cudaArchList` argument that allows users to test on the latest nvidia hardware without waiting for this package.
- adds mkl support to magma-2.5.0
- adds cuda support to openMPI
- adds openMPI support
- allows users to build NamedTensors
- allows users to get access to generated binaries from build process
- adds `.dev` output which gives top-level access to generated C/C++ code (libtorch).
- adds darwin support if `cudaSupport` is disabled.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

tbenst commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,6 +67,12 @@ in buildPythonPackage rec {
 
+  USE_FBGEMM = 0; # this can't build because of CMAKE downloads
+  USE_MKLDNN = 0; # mkl with cuda is broken
+  USE_NCCL = 0; # multigpu looks broken broken
```
Did you try including [NCCL](https://github.com/NixOS/nixpkgs/blob/5d86f9a9ce86be51f20d213a43e96f8694f08169/pkgs/development/libraries/science/math/nccl/default.nix)?

stites replied:
Yes, unfortunately.

stites replied:
Maybe this is because nvlink is attempting to perform checks in the GPU runtime?

tbenst commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -36,14 +39,6 @@ in buildPythonPackage rec {
     sha256 = "076cpbig4sywn9vv674c0xdg832sdrd5pk1d0725pjkm436kpvlm";
```
Is this SHA correct? Looks the same to me as 1.0.0?

stites replied:
ah! Yes, I'm primarily working off of https://github.com/stites/libtorch-nix and missed this when transferring over details. Good catch -- I'm a bit surprised this slipped through since I was building off of my fork (using the command from the PR message).

tbenst commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,6 +67,12 @@ in buildPythonPackage rec {
   PYTORCH_BUILD_VERSION = version;
   PYTORCH_BUILD_NUMBER = 0;
 
+  USE_FBGEMM = 0; # this can't build because of CMAKE downloads
```
Worth reviewing the instructions at https://github.com/pytorch/FBGEMM. Looks like we can avoid downloads by setting ASMJIT_SRC_DIR, CPUINFO_SRC_DIR, GOOGLETEST_SOURCE_DIR.

stites replied:
So I did set these as environment variables, using pointers from your gist: https://gist.github.com/tbenst/15495f97ffe2a71dc18c49e247af8f51#file-pytorch_11-nix-L62

Unfortunately, these didn't seem to propagate through to CMake. I'll do a round-two later.

tbenst reviewed (commented):
Thanks for working on this! Hopefully we can get this as feature complete as possible.

stites commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,6 +67,12 @@ in buildPythonPackage rec {
   PYTORCH_BUILD_NUMBER = 0;
 
+  USE_FBGEMM = 0; # this can't build because of CMAKE downloads
+  USE_MKLDNN = 0; # mkl with cuda is broken
```
MKL with cuda is _not_ broken, actually -- just MKLDNN. The problem seems to be a warning which is treated as an error:
```
/build/source/third_party/ideep/mkl-dnn/src/cpu/jit_uni_reorder_utils.cpp: In function ‘void mkldnn::impl::cpu::tr::prb_normalize(mkldnn::impl::cpu::tr::prb_t&)’:
/build/source/third_party/ideep/mkl-dnn/src/cpu/jit_uni_reorder_utils.cpp:273:29: error: array subscript is above array bounds [-Werror=array-bounds]
                 || p.nodes[j].os < p.nodes[min_pos].os
                    ~~~~~~~~~^
/build/source/third_party/ideep/mkl-dnn/src/cpu/jit_uni_reorder_utils.cpp:276:37: error: array subscript is above array bounds [-Werror=array-bounds]
                         && p.nodes[j].n < p.nodes[min_pos].n);
                            ~~~~~~~~~^
cc1plus: all warnings being treated as errors
make[2]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/build.make:1272: third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/cpu/jit_uni_reorder_utils.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

stites replied:
Found the issue: https://github.com/pytorch/pytorch/issues/22346

stites replied:
I missed the workaround at https://github.com/NixOS/nixpkgs/blob/fb76f093c9ab204fbec9dc9c56cb9b4a81f43bc7/pkgs/development/python-modules/pytorch/default.nix#L79

I'll be adding a `mklSupport ? false` which will swap out the numpy BLAS implementation so that this is more explicit. The working code is similar to `my_magma`:

```
  my_numpy = if mklSupport && numpy.blasImplementation != "mkl" then numpy.override { blas = mkl; } else numpy;
```

stites commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,6 +67,12 @@ in buildPythonPackage rec {
+  USE_MKLDNN = 0; # mkl with cuda is broken
+  USE_NCCL = 0; # multigpu looks broken broken
+  # on deck
+  USE_DISTRIBUTED = 0; # this might work
```
Distributed fails because of the testing suite. I'll rebuild and pull out the stack trace.

stites replied:
Seems to have resolved itself

stites commented:
I've just rebased this on top of master, which means I need to cherry-pick and remove the #61347 commit (65494f2) before this is finalized. I don't think this should be a problem since I believe everything needs to be squashed anyways.

I've updated the PR with the following:
- NCCL support is back
- DISTRIBUTED is turned back on
- OpenMPI support is now available (I needed to get the latest master so that I could update it's `default.nix` with CUDA support). I am tempted to change `openMPISupport` to `distributedSupport`
- SHA has been corrected
- mkl support is synced with numpy and correctly builds MKL_DNN
- ninja and setuptools have been added to the build (pytorch makes them optional, but I think this speeds up the build a bit).

Things that still need to be worked on:
- FBGEMM support

stites commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -1,17 +1,24 @@
   };
+  my_magma = magma.override {cudatoolkit = cudatoolkit;};
+  my_numpy = if mklSupport && numpy.blasImplementation != "mkl" then numpy.override { blas = mkl; } else numpy;
+  my_openmpi = if openMPISupport then openmpi.override { inherit cudaSupport cudatoolkit; } else openmpi;
```
This can be moved to a new PR after everything has been tested.

stites commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -25,25 +32,17 @@ let
+    sha256 = "1rckv7m3h04mgf7p61rmabszgxh5110ly6qq0qnp28vg7mckvgkh";
   };
 
-  patches =
```
I believe this patch can be removed since it was patching with a v1.1.0 commit. I can verify on a single-gpu instance later.

stites commented:
FBGEMM works in https://github.com/stites/libtorch-nix/blob/master/deps/fbgemm.nix but, unfortunately, setup.py seems to hide arguments passed into `cmakeFlags` from cmake.

stites commented:
I asked around on the pytorch slack and the 1.1.1 release is going to be cut this week, so I think waiting is the best course of action.

For posterity, if anyone is trying to build pytorch-1.1.0, this PR is fully-functional without FBGEMM -- which provides optional, low-precision general matrix multiplication for small batch sizes, various MM-based optimizations, and fused operators. If you are working off of an x86 CPU backend, however, having FBGEMM might be a bit of a nessecity.

stites commented:
https://github.com/pytorch/pytorch/releases/tag/v1.2.0

Looks like they skipped the 1.1.1 release. I'll update this PR for pytorch-1.2.0.

stites commented:
Still working on this, slowly. I am also going to set `platforms = with lib.platforms; [linux] ++ lib.optionals (!cudaSupport) [darwin];` unless there are any objections.

stites commented:
I made an update-worthy breakthrough for pytorch-1.2.0. The errors I saw weren't with NCCL, but with the 3.0 CUDA architectures, documented in https://github.com/pytorch/pytorch/issues/23573. I'm currently trying to figure out how to polish this up and did notice that cmake was throwing this warning:

```
CMake Warning at cmake/public/utils.cmake:172 (message):
  In the future we will require one to explicitly pass TORCH_CUDA_ARCH_LIST
  to cmake instead of implicitly setting it as an env variable.  This will
  become a FATAL_ERROR in future version of pytorch.
```

Which is quite troubling. So far, we have been using `buildPythonPackage` to install pytorch, but I think we could can use the [*Adjust Build Options*](https://github.com/pytorch/pytorch/tree/v1.2.0#adjust-build-options-optional) instructions for more flexibility around cmake (so this would have cracked the FBGEMM problem in 1.1.0), as well as future proof the pytorch derivation.

I'll be looking at adjusting `buildPythonPackage` next.

---

*NOTE:* Because of sandboxing, the build can't auto-detect the hardware's cuda architecture, so there is also now a problem around new architectures not being supported until explicitly added to this derivation.

stites commented:
`buildPythonPackage` is actually a bit too much for me right now. I've bumped https://github.com/stites/pytorch-world to build for all variants of pytorch-1.2 (tested on python-3.6) -- cachix binaries are also available for testing.

I've added the above note about CMake in comments and will update this PR this week -- I've also added myself as a maintainer for this package.

stites commented:
I've gone ahead and pushed up the latest changes, pytorch-1.2.0 is fully functional! Note that this is still blocked by #61347 and I can squash the commits if desired, but this PR now touches openmpi (adding cuda support to it) and magma (adding mkl support to it).

Final notes:
- I've added a `cudaArchList` because of notes above regarding sandboxing -- this should allow users to test on the latest gpu hardware without waiting for this package. The arch lists haven't been vetted and I just notice a minute ago that cuda-9 was broken, so I've split it up into -9 and -10 versions. This logic feels dirty.
- darwin is now supported if you turn off cudaSupport (but it's still untested).
- I've added `dev` outputs -- this is a quick hack that gives users like hasktorch (@tscholak) the generated C/C++ backend (thanks @cleverca22!). I haven't tested this, but you have to use python36Packages.callPackage ([example here](https://github.com/stites/pytorch-world/blob/master/release.nix#L22)). There's probably a better way to do this (possibly promoting pytorch to top-level and using the "Adjust Build Options" instructions).
- Because the original derivation mutates numpy via passthrough, it is very easy to get a "duplicate numpy package found." I am thinking it would be safer to assert `numpy.blasImplementation == "mkl"` if mklSupport is present (but I also used the same trick for magma-mkl support and openmpi-cuda support, so ¯\_(ツ)_/¯. 

This is ready for review. Again, not sure if I need to squash all of this or split up the PR into parts. Because I just found out how to properly support cuda-9, [cachix binaries are pending](https://pytorch-world.cachix.org/) -- I'll comment one last time to ping folks who want to test-drive this (although only the vanilla cuda-9 build is in there).

stites commented:
Cachix is ready. I'm unaware of how cachix works but if you redo the steps I did with cachix turned on, things might run a little faster: https://github.com/stites/pytorch-world#cachix

wavewave commented:
Great work!

wavewave commented:
Don't we need to handle the following issue? 
https://discuss.pytorch.org/t/issues-linking-with-libtorch-c-11-abi/29510

I am now making this nix-shell environment (https://github.com/wavewave/hasktorch/blob/8c01de9143405f59ae77090bf31d44376b8ed483/shell.nix) to compile hasktorch, and `cabal new-build libtorch-ffi` worked, but `cabal new-build hasktorch` ended up with this error: 
```
[nix-shell:~/repo/srcc/hasktorch]$ cabal new-build hasktorch
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - hasktorch-0.2.0.0 (lib) (first run)
Preprocessing library for hasktorch-0.2.0.0..
Building library for hasktorch-0.2.0.0..
<command line>: can't load .so/.DLL for: /home/wavewave/repo/srcc/hasktorch/dist-newstyle/build/x86_64-linux/ghc-8.6.5/libtorch-ffi-1.1.0.0/build/libHSlibtorch-ffi-1.1.0.0-inplace-ghc8.6.5.so (/home/wavewave/repo/srcc/hasktorch/dist-newstyle/build/x86_64-linux/ghc-8.6.5/libtorch-ffi-1.1.0.0/build/libHSlibtorch-ffi-1.1.0.0-inplace-ghc8.6.5.so: undefined symbol: _ZN3c105ErrorC2ENS_14SourceLocationERKSs)

$ [nix-shell:~/repo/srcc/hasktorch]$ c++filt _ZN3c105ErrorC2ENS_14SourceLocationERKSs
c10::Error::Error(c10::SourceLocation, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
```  
which seems to be related to C++11 ABI.

wavewave commented:
ah. never mind. i found this hard-coded flag in libtorch-ffi: https://github.com/hasktorch/hasktorch/blob/2f2a5e9e32389114d5768f8f934c45c19623baf3/libtorch-ffi/libtorch-ffi.cabal#L74-L78

junjihashimoto commented:
You use the libtorch-library not using C++11ABI.
Because the error includes `std::basic_string`.
> c10::Error::Error(c10::SourceLocation, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)

If you use the libtorch-library using C++11ABI, the error may includes `std::__cxx11::basic_string`.

tscholak commented:
has anyone except me tried building this on a Mac?
as @stites knows, I have been trying to do this for some time now, and I am running into a weird `BUS ERROR` during testing.
I'm running this again as I am writing this comment, and I will post the full error message shortly.

teh commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,38 +136,55 @@ in buildPythonPackage rec {
+  checkPhase = "${cudaStubEnv}python test/run_test.py"
+    + " --exclude utils" # utils requires git, which is not allowed in the check phase
+
+    # Other tests which have been disabled in previous nix derivations of pytorch.
```
If these tests pass now I thin you can remove?

stites replied:
Sure, but maybe a better way to do this would be to leave these and pass along a `doCheck = false`? Would that work? I'm still trying to figure out the python-nix-ecosystem. I do want to make sure that people can verify that everything is functional on their box if desired.

stites replied:
I've added `doCheck=false;` above. Let me know if this isn't enough (maybe I misunderstand doCheck)

tscholak replied:
little OT: I suspect that with `doCheck=false;` the build will succeed on Darwin but then will fail later at runtime ;)

tscholak replied:
@stites I merged this commit into my `nix` branch, and it had the desired (?) effect of skipping all tests for pytorch.

teh reviewed (approved):
Thank you, fantastic work!

stites commented:
@tscholak were you able to figure out the build issues on OSX, since you have been able to make forward progress on https://github.com/hasktorch/hasktorch/pull/165 -- right? Double-checking on wikipedia, this bus error is hardware-related, something to do with memory access, perhaps?

tscholak commented:
@stites I've been pushing nix for hasktorch on three different platforms in parallel: Darwin, docker, and nixos native. 
1. On Darwin, I never got past that SIGBUS issue.
2. On docker, I different fatal errors during testing, referenced elsewhere.
3. On nixos native, I was able to develop a suite of working nix derivations for hasktorch and its examples.
I didn't try anything with CUDA yet. The only CUDA-enabled hardware that is available to me right now is running ubuntu, so I will need to use nvidia-docker.

tscholak commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -72,38 +136,57 @@ in buildPythonPackage rec {
+    mkdir $dev
+    cp -r $out/${python.sitePackages}/torch/lib     $dev/lib
+    cp -r $out/${python.sitePackages}/torch/include $dev/include
   '';
```
Hi @stites!
I'm running into a new problem on `Darwin` when building hasktorch's `libtorch-ffi`:
```
Preprocessing test suite 'spec' for libtorch-ffi-1.1.0.0..
Building test suite 'spec' for libtorch-ffi-1.1.0.0..
[1 of 5] Compiling BackwardSpec     ( test/BackwardSpec.hs, dist/build/spec/spec-tmp/BackwardSpec.o ) [flags changed]
[2 of 5] Compiling BasicSpec        ( test/BasicSpec.hs, dist/build/spec/spec-tmp/BasicSpec.o ) [flags changed]
[3 of 5] Compiling CudaSpec         ( test/CudaSpec.hs, dist/build/spec/spec-tmp/CudaSpec.o ) [flags changed]
[4 of 5] Compiling MemorySpec       ( test/MemorySpec.hs, dist/build/spec/spec-tmp/MemorySpec.o ) [flags changed]
[5 of 5] Compiling Main             ( test/Spec.hs, dist/build/spec/spec-tmp/Main.o ) [flags changed]
Linking dist/build/spec/spec ...
running tests
Running 1 test suites...
Test suite spec: RUNNING...
dyld: Library not loaded: @rpath/libc10.dylib
  Referenced from: /private/tmp/nix-build-libtorch-ffi-1.1.0.0.drv-0/libtorch-ffi/dist/build/spec/spec
  Reason: image not found
Test suite spec: FAIL
Test suite logged to: dist/test/libtorch-ffi-1.1.0.0-spec.log
0 of 1 test suites (0 of 1 test cases) passed.
builder for '/nix/store/s6av6bgxnsf53n06ryxmc9ksajbwasvi-libtorch-ffi-1.1.0.0.drv' failed with exit code 1
error: build of '/nix/store/s6av6bgxnsf53n06ryxmc9ksajbwasvi-libtorch-ffi-1.1.0.0.drv' failed
```
I tried a bunch of things short of injecting the torch libs via `LD_LIBRARY_PATH`, but `libc10.dylib` remains to be the problem. I suspect it has to do with the fact that macOS has a weird lookup mechanism for this dylibs that needs additional build steps to work properly.
For instance, consider these lines in the MKL derivation: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/science/math/mkl/default.nix#L61-L71
As you can see, `install_name_tool` is applied to every dylib in the package.
See https://stackoverflow.com/questions/33991581/install-name-tool-to-update-a-executable-to-search-for-dylib-in-mac-os-x for more information.

stites commented:
After discussion with @tscholak, I've added him to the maintainer list for pytorch as the darwin-build-expert. I believe this is now ready for final review, once #61347 is merged, I will cherry-pick out all magma-related commits, squash, and rebase.

jmillerpdt commented on pkgs/development/libraries/science/math/magma/default.nix:
```diff
@@ -1,25 +1,50 @@
-{ stdenv, fetchurl, cmake, gfortran, cudatoolkit, libpthreadstubs, liblapack }:
+{ stdenv, fetchurl, cmake, gfortran, cudatoolkit, libpthreadstubs, liblapack
+, mklSupport ? false, mkl ? null
+}:
```
You could also pick up mkl-dnn (USE_MKLDNN=ON) and protobuf (BUILD_CUSTOM_PROTOBUF=0) directly from Nix, which should reduce build time and improve interoperability between it and other Nix built libraries.

stites replied:
For magma? Magma is now merged and these diffs will be removed from the PR.
---
For pytorch:

Protobuf requires more digging to see how to correctly hook it into pytorch's build process since environment variables are masked. Really this means that I need to figure out how to build the nix derivation with https://github.com/pytorch/pytorch#adjust-build-options-optional. I'm on board, but I think this should be another PR.

MKLDNN is not yet in nix, the open PR is here: https://github.com/NixOS/nixpkgs/pull/68014. It's not clear that `USE_MKLDNN` does the thing you want, however, since I believe MKLDNN is built by default if mkl is present (hence the [mkldnn hack](https://github.com/NixOS/nixpkgs/blob/187025529c4edeb42ed7e688675083097ac05ff5/pkgs/development/python-modules/pytorch/default.nix#L142)). `USE_*` flags generally just toggle, from what I've seen, and I can't find anything in https://github.com/pytorch/pytorch/blob/v1.2.0/setup.py which would allow us to set the path. Using CMake directly might, but then I'd have to figure out how to partially run `buildPythonPackage` and then run `ccmake`. w.r.t. mkldnn I'm also a little concerned in version-matching if there isn't a clean configuration in the setup.py.

stites replied:
Followup: I tried both of these things and couldn't get either package to be hooked up correctly on a first pass.

jmillerpdt replied:
@stites Thanks.  Once MKL-DNN is available, I'll try to get a follow-up PR to you for these.  Appreciate you taking a look.

stites replied:
Please do! I would actually encourage you to try it out using https://github.com/stites/pytorch-world, which is where I am drafting all the work that goes in here.

prusnak commented:
@stites seems that https://github.com/NixOS/nixpkgs/pull/61347 is merged now! :tada:

stites commented:
This is ready for final review.

elohmeier commented:
This will help us with #68361. Thanks! :-)

teh commented:
I think probably out of scope for this PR but when I tested the tensorboard integration I get this:

```
nix-shell -p '(python3.withPackages (ps: [ps.pytorch ps.tensorflow-tensorboard]))'
python3
>>> import torch.utils.tensorboard
...
    from .writer import FileWriter, SummaryWriter  # noqa F401
  File "/nix/store/2yzgdpfa54ckx9x662l8rq1x50d6ibgd-python3-3.7.4-env/lib/python3.7/site-packages/torch/utils/tensorboard/writer.py", line 18, in <module>
    from ._convert_np import make_np
  File "/nix/store/2yzgdpfa54ckx9x662l8rq1x50d6ibgd-python3-3.7.4-env/lib/python3.7/site-packages/torch/utils/tensorboard/_convert_np.py", line 12, in <module>
    from caffe2.python import workspace
  File "/nix/store/2yzgdpfa54ckx9x662l8rq1x50d6ibgd-python3-3.7.4-env/lib/python3.7/site-packages/caffe2/python/workspace.py", line 15, in <module>
    from past.builtins import basestring
ModuleNotFoundError: No module named 'past'
```

I suspect it's this dependency https://pypi.org/project/past/ which hasn't been packaged for nixpkgs yet AFAICT.

stites commented:
No! I didn't notice the conflicts. I'll have to get to that next week.

@teh I haven't checked but tensorboard integration should already be packaged with the latest release of PyTorch. You would need to install tenserflow-board separately, but only for the binary.

stites commented:
@teh -- looks like it's missing https://github.com/PythonCharmers/python-future, also six from browsing the tensorboard source. Verifying the build now.

stites commented:
rebased the commit over master, double-checked that openmpi is still correct, added extra python dependencies for tensorboard (but there is still an issue of `tensorflow-tensorboard` not having the required packages). Diff comes from https://github.com/stites/pytorch-world/commit/0e59fe263f21582e98084748e42f51f8fa67bfdb

Ready for another round of review.

jethrokuan commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -1,17 +1,88 @@
+  ninja,
+
+  # dependencies for torch.utils.tensorboard
+  tensorboardSupport ? true, pillow, six, future, tensorflow-tensorboard,
```
I suppose tensorflow is missing here?

```
error: undefined variable 'tensorflow' at /home/jethro/nix-config/nixpkgs/pkgs/development/python-modules/pytorch/default.nix:173:83
```

stites replied:
Oops! tensorflow should have been removed from [line 173](https://github.com/NixOS/nixpkgs/pull/65041/files#diff-84b29016f8c93846ea69f3885d709770R173).

jonringer commented:
I know this is probably wearing on you, but do you mind rebasing off master? :)

stites commented:
Rebased -- I assume that `cmakeUseConfigure = false;` just disables cmake configure, which is fine since pytorch is configured through `setup.py`. This might need to be disabled in the future if we start to use the [_Adjust Build Options_](https://github.com/pytorch/pytorch/tree/v1.2.0#adjust-build-options-optional) build instructions. I can't find any documentation on `cmakeUseConfigure`, though -- so this is all based on assumption.

The pytorch-world build and cachix push have been retriggered via https://github.com/stites/pytorch-world/commit/0f3af8e5160d15643186df2d3745dbdfb2c612e3. It would be nice to get this merged in, though.

jonringer commented:
another options is to use `dontUseCmakeBuildDir = true;`. The main problem for a lot of python packages is that by default it will do a normal cmake work flow of `mkdir build; cd build; cmake ..`. But this will drop you into a directory in which setup.py isn't present.

jonringer commented:
Personally I gave up on looking for documentation (probably a bad habit), and just go to the source... The logic you're looking for is contained here https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/cmake/setup-hook.sh

jonringer commented on pkgs/development/python-modules/pytorch/default.nix:
```diff
@@ -73,38 +143,83 @@ in buildPythonPackage rec {
   propagatedBuildInputs = [
     cffi
-    numpy
+    my_numpy
```
Generally you want to avoid propagating different versions of the same library, there might be a scenario where a user will have a pytorch + pyproj environment in which they both use numpy. The numpy dependency from both of those packages will be put on PYTHONPATH, and you can't guarentee the order in which they get traversed. So you may end up in a scenario where you dont have mkl support.

If this was an application, you could override the python package set, and then your dependencies should inherit the overrided packages transitively.

jonringer replied:
@FRidh correct me if I'm wrong, because I want to be :)

jmillerpdt replied:
@stites are you just trying to work around the fact tensorflow has not yet picked up the ability to compile with MKL? (#69454)

stites replied:
This was a pattern introduced prior to this PR -- I definitely feel this pain when working on personal projects. Before replicating this pattern in both magma (for mkl) and openMPI (for cuda) I recall reviewing a PR where someone said this was the right course of action. I'd be happy to change it, though, as it did cause some headaches in the beginning.

Ideally, I'd like to trigger a warning that says something like "building with mklSupport, but magma and/or numpy have not been built with mkl." I'm not sure how to do this, though... Maybe just a loud echo in the build phase? openMPISupport can be added to the assert statements.

tbenst replied:
yeah, sorry I introduced this pattern as a newbie crutch :grimacing:. I think the better place to do this is with overrides on `callPackage`? E.g. https://github.com/NixOS/nixpkgs/blob/8cd5aa8a82367eeedbd0261f6d040a106e2a844a/pkgs/top-level/all-packages.nix#L7565

jonringer replied:
the main problem is that when this gets built by hydra, you will just get a cache hit, and pull down the closure. You won't see any build output.

jonringer replied:
cross-referencing python scoping issues #67422

stites replied:
Rebuilding pytorch-world with changes now. I'm going to assert that the same mkl/cudatoolkit is sync'd across all dependencies (note the enforced mkl instead of the verbose build output). This mimics how the derivation works now, but requires the user to supply correct dependencies. 

Asserts look something like:
```
assert !cudaSupport || lib.lists.any (inp: inp == cudatoolkit) magma.buildInputs;
```

vladmaraev commented:
I tried to compile this, but I get the following message:
> error: Package ‘python3.7-pytorch-1.2.0’ in /home/vlad/nixpkgs/pkgs/development/python-modules/pytorch/default.nix:219 is not supported on ‘x86_64-linux’, refusing to evaluate.

stites commented:
@vladmaraev -- I'm not clear on the details to debug your machine. I assume that [line 222](https://github.com/NixOS/nixpkgs/pull/65041/files#diff-84b29016f8c93846ea69f3885d709770R222), `platforms  = with lib.platforms; [ linux ] ++ lib.optionals (!cudaSupport) [ darwin ];`, would resolve your problem. [Line 219 is `meta.description`](https://github.com/NixOS/nixpkgs/pull/65041/files#diff-84b29016f8c93846ea69f3885d709770R219), so it's possible you are on an older commit.

Can you try something like the following:
```
$ nix-shell \
  -I nixpkgs=/home/stites/git/nixpkgs \
  -p '(python37.withPackages (ps: [ps.pytorch]))' \
  --run 'python -c "import torch; print(torch.__version__); import platform; print(platform.platform())"'
```

On NixOS I get:
```
1.2.0
Linux-4.19.68-x86_64-with
```

While I think the likely scenario is a wrong `nixpkgs`, it's still curious that your derivation refuses to evaluate (since it should be building prior to this PR). Can you pass along your build details if this persists?

vladmaraev commented:
@stites my line 222 looks exactly like you mentioned. I only can build it when I change it to the following: `platforms   = lib.platforms.linux;`
Then output is:
```
1.2.0
Linux-5.0.17-200.fc29.x86_64-x86_64-with-fedora-29-Twenty_Nine
```

But I can't build it with CUDA support: (`ps.pytorchWithCuda`):
```
make: *** [Makefile:5073: magma] Error 2
builder for '/nix/store/jnwxbh823cb7wd0wv9bgivs7l1f9s3nr-magma-2.5.0.drv' failed with exit code 2
cannot build derivation '/nix/store/0hsymlzk898j3gk6r2h00jrr4w1gqlin-python3.7-pytorch-1.2.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/j42qvpk5hjbamwnxavvapjhsggmdrsyh-python3-3.7.4-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/j42qvpk5hjbamwnxavvapjhsggmdrsyh-python3-3.7.4-env.drv' failed
```

stites commented:
Hmm... @vladmaraev, Can you try to build with `platforms = with lib.platforms; [ linux darwin];`? If that works, I'll add an assert to ensure that darwin doesn't run with cuda support.

Regarding `pytorchWithCuda`, I'll double-check this and the magma derivation. Currently, I am rebuilding the 17 pytorch-world derivations without the mutation pattern (from https://github.com/NixOS/nixpkgs/pull/65041#discussion_r330674100

vladmaraev commented:
@stites  No, this also doesn't work. But this does: `platforms = with lib.platforms;  linux ++ darwin;`

stites commented:
Very bizzare -- it seems like this shouldn't be building on my machine at all! both linux and darwin are singleton lists, so the types shouldn't match up.

@vladmaraev, can you try to build with `platforms = with lib.platforms; linux ++ lib.optionals (!cudaSupport) darwin;`? I think it would be better to block darwin-cuda support here instead of in a top-level assert statement.

vladmaraev commented:
> can you try to build with `platforms = with lib.platforms; linux ++ lib.optionals (!cudaSupport) darwin;`? 
yes, it works.

jonringer commented:
`platforms = with lib.platforms; [ linux darwin];` doesnt work because it would be type `[[ string ]]` when you want type `[ string ]`, `linux ++ ...` is indeed, what you want :)

stites commented:
Okay, I've pushed up the assert statements to resolve @jonringer's concerns, fixed the type of `meta.platforms`, and also added the same `preConfigure` hook to magma to ensure that it doesn't fail for `pytorchWithCuda` (ie -- [I've made sure magma builds correctly with cudatoolkit9.2](https://github.com/NixOS/nixpkgs/pull/65041/commits/6213c2821b0ec03e3f792c0a0a320f31edc4d583#diff-c050e4bcc21558e68cf9774a3123d8aeR27-R30)). @vladmaraev can you confirm one last time this is working for you (as you are currently having the most trouble).

I have tested with the following:
```
$ nix-shell \
  -I nixpkgs=/home/stites/git/nixpkgs \
  -p '(python37.withPackages (ps: [ps.pytorch]))' \
  --run 'python -c "import torch; print(torch.__version__); import platform; print(platform.platform())"'
```
and
```
$ nix-shell \
  -I nixpkgs=/home/stites/git/nixpkgs \
  -p '(python37.withPackages (ps: [ps.pytorchWithCuda]))' \
  --run 'python -c "import torch; print(torch.__version__); import platform; print(platform.platform())"'
```

I am also rebuilding the latest pytorch-world -- all derivations build with the new asserts and I'm just doing one final pass with the new magma script. I've been building with pytorch-3.6+cuda-9.0 and pytorch-3.6+cuda10.0 -- I've added a pytorch-3.7+cuda9.2 build as well which has already passed.

vladmaraev commented:
I run:
```
nix-shell   -I nixpkgs=/home/vlad/nixpkgs \
   -p '(python37.withPackages (ps: [ps.pytorchWithCuda]))' \
  --run 'python -c "import torch; print(torch.__version__); import platform; print(platform.platform()); print(torch.cuda.is_available());";'
1.2.0
Linux-5.0.17-200.fc29.x86_64-x86_64-with-fedora-29-Twenty_Nine
False
```

So it builds but CUDA is not available.

stites commented:
@vladmaraev This might be an issue with your box's cuda configuration and how it interacts with nix. Unfortunately, I am not able to replicate this on my box.

vladmaraev commented:
@stites I will take a look what can be done regarding interaction of nix and cuda. Just as a note, this commit https://github.com/NixOS/nixpkgs/commit/02648a6cc1816b7f0156a5a2b9dcee2c0fbd37cc builds fine and shows `torch.cuda.is_available()` as True, but when I am trying to train my model I get CUDA unknown error.

stites commented:
Can you link a gist/repo for me to run? CUDA-based models have been running fine on my end (so far I'm just running some probabilistic programming frameworks and haven't used the full extent of pytorch).

Also, I noticed your `platform.platform()` call references fedora -- are you running nix-on-fedora? If you aren't on NixOS, I would also verify that you are building with system-level dependencies in your shell. Perhaps double-checking https://nixos.wiki/wiki/Nvidia#CUDA and re-adjusting your PATH/LD_LIBRARY_PATH/CUDA_PATH, might be necessary.

vladmaraev commented:
@stites I forgot to add `LD_PRELOAD=/lib64/libcuda.so.1:/lib64/libnvidia-fatbinaryloader.so.418.74` before running python, and after adding this line it works, so I can run the model (hurray!). Thanks a lot.

The example I am running is not publicly available, but here is the minimal example that will work for this branch but won't work for https://github.com/NixOS/nixpkgs/commit/02648a6cc1816b7f0156a5a2b9dcee2c0fbd37cc:

```
torch.as_tensor([1,1], device='cuda')
```

stites commented:
@vladmaraev glad to hear things are working! Is `torch.as_tensor([1,1], device='cuda')` line no longer a problem? I am unable to replicate this so hopefully adjusting LD_PRELOAD has fixed this.

vladmaraev commented:
@stites Yes, this is not a problem anymore. Now I having an issue with (supposedly) convolution in pytorch: I get Segmentation Fault. Now I am trying to rebuild it with `cudatoolkit_10` instead of version 9. I will provide you with detail in case if it will not work.

vladmaraev commented:
It works with `cudatoolkit_10` now, probably there were some changes in CUDA regarding convolutional layers. Would you mind updating it: `all-packages.nix#L2493` and `all-packages.nix#L2507`?

stites commented:
Yeah, I do recall pytorch doesn't play nice with cuda 9.2 and only added support for it much later (and after cuda-10). I would assume that these are fixed, however.

I can change make this change, but because cuda 9.2 is the default of nixos I am concerned that defaulting support to 10.0 might have adverse effects. Can someone deeper into nix-gpu knowledge comment on this? Another alternative is to add a new `pytorchWithCuda_10` script and mark `pytorchWithCuda` as broken (but keep it around for bookkeeping). Thoughts?

jyp commented:
@stites There should be no problem instantiating pytorch locally with cuda/cudnn 10.0. This is what we have been doing for years with tensorflow. We have:

```
  tensorflow-bin = callPackage ../development/python-modules/tensorflow/bin.nix {
    cudaSupport = pkgs.config.cudaSupport or false;
    inherit (pkgs.linuxPackages) nvidia_x11;
    cudatoolkit = pkgs.cudatoolkit_10;
    cudnn = pkgs.cudnn_cudatoolkit_10;
  };
```

I imagine exactly the same pattern will work for pytorch.

jyp commented:
After checking the commit I see that `openmpi` and `magma` also depend on cuda. It stands to reason that these low-level libraries should be using the default version of cuda. Thus I withdraw my previous comment and simply recommend to merge the PR as it stands. @vladmaraev should be able to override his version of cuda locally until nixpkgs moves to cuda 10 (hopefully soon).

jonringer commented:
@stites do you mind rebasing? looks like there was a tree wide "mark broken builds as broken" change.

teh commented:
This passes my internal test cases so LGTM from me. I think we should merge straight after rebase unless anyone else objects.

stites commented:
Rebased! After this PR, I'll start on 1.3.0

teh commented:
@jonringer good to merge?

jonringer commented:
even with my 3900X, this takes a while to build maxing out all 24 threads >.>

jonringer reviewed (commented):
is caffe2 supposed to be bundled with pytorch?

```
[nix-shell:/home/jon/.cache/nix-review/pr-65041-1]$ tree -L 4 ./results/python37Packages.pytorchWithCuda/
./results/python37Packages.pytorchWithCuda/
├── bin
│   ├── convert-caffe2-to-onnx
│   └── convert-onnx-to-caffe2
└── lib
    └── python3.7
        └── site-packages
            ├── caffe2
            ├── torch
            └── torch-1.2.0.dist-info
```
if so, please add click to the propagatedBuildInputs:
```
[nix-shell:/home/jon/.cache/nix-review/pr-65041-1]$ ./results/python37Packages.pytorchWithCuda/bin/convert-caffe2-to-onnx --help
Traceback (most recent call last):
  File "/nix/store/cbrg1ldydrl4wb6v4xpkkfh2jbmhwaiv-python3.7-pytorch-1.2.0/bin/.convert-caffe2-to-onnx-wrapped", line 7, in <module>
    from caffe2.python.onnx.bin.conversion import caffe2_to_onnx
  File "/nix/store/5ck5ncbzq8pr9aynzyp3j2y0bhk0v42i-python3.7-pytorch-1.2.0/lib/python3.7/site-packages/caffe2/python/onnx/bin/conversion.py", line 11, in <module>
    import click
ModuleNotFoundError: No module named 'click'
```

jonringer commented:
closes #69424

stites commented:
caffe2 is used for the "production-ready" features of pytorch.

I've added click to the propagatedBuildInputs. My ML box is currently down (just moved offices), so this is a bit of an optimistic push.

jonringer commented:
I'm building and testing now. If it seems to work in the trivial case, I'll merge :)

jonringer commented:
@GrahamcOfBorg eval

jonringer commented:
not sure how much you use onnx, but I added a cpu-only version to nixpkgs #67542

jonringer reviewed (approved):
`nix-review` passes on NixOS (python38Packages.pytest is broken)
diff LGTM

```
[nix-shell:/home/jon/.cache/nix-review/pr-65041-2]$ python
Python 3.7.4 (default, Jul  8 2019, 18:31:06)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.2.0'
```
```
[3 built (2 failed), 0.0 MiB DL]
error: build of '/nix/store/zqv58qrag64v4db7ljcmk62r3w4vjjd7-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/65041
3 package failed to build:
python38Packages.pytorch python38Packages.pytorchWithCuda python38Packages.torchvision

4 package were build:
magma python37Packages.pytorch python37Packages.pytorchWithCuda python37Packages.torchvision
```
the caffe2 executables still need some love, but I'll do that in a different PR seeing as this one has been going on for close to 3 months

jonringer reviewed (changes requested):
ooops, to conform to contirbuting.md, please break out the work into the following commits:

```
maintainers: add tscholak
openmpi: enable cuda support
magma: use cudatoolkit
python3Packages.pytorch: 1.0.0 -> 1.2.0
```

stites committed:
maintainers: add tscholak

stites committed:
openmpi: enable cuda support

stites committed:
magma: use cudatoolkit

stites committed:
python3Packages.pytorch: 1.0.0 -> 1.2.0

stites commented:
quick ping to mention that this change was made!

jonringer commented:
I might wait for #71780 to get in, as it has significant improvements to python38. And it's really hard to not have 480 commits bit rot fast

jonringer commented:
fridh said he would de-conflict the merge at a later time.

stites commented:
Heads-up: an initial build of pytorch-1.3.0 succeeds on python3.7 with and without cuda. Test are failing, do to some dependencies issues. The first example is that there are now a first-class tensorboard test suite.

I'm also about to be very busy for the next week -- expect a PR shortly after.

jonringer commented:
no sweat, take your time man. We appreciate your effort :)

stites commented:
vanilla pytorch-1.3.0 builds with a bumped tensorboard-2.0 (incoming PR). Triggering a full dependency grid-sweep now with tests

danieldk commented:
I think 2947c23d21f4b9b68558620113f12b182ea78e26 should also be backported to 19.09? `magma` builds with CUDA on 19.09 fail with:

```
In file included from /nix/store/lzwx11791zlw77swqakrzjrkm5mz84cq-cudatoolkit-10.0.130/include/cuda_runtime.h:83,
                 from <command-line>:
/nix/store/lzwx11791zlw77swqakrzjrkm5mz84cq-cudatoolkit-10.0.130/include/crt/host_config.h:129:2: error: #error -- unsupported GNU version! gcc versions later than 7 are not supported!
 #error -- unsupported GNU version! gcc versions later than 7 are not supported!
  ^~~~~
```
